### PR TITLE
Add CI for serial-keel

### LINKED ISSUE #7: Add CI
Perhaps start by copying this: https://github.com/torsteingrindvik/bevy-vfx-bag/blob/main/.github/workflows/basics.yml

At least

- Build
- cargo fmt
- cargo clippy
- cargo doc

Also we'd like to run the tests. That might or might not be more difficult, but

- cargo test --all-features

would be great.

For python, consider a linter and/or formatting checker.

### PR DESCRIPTION
**Pull request by stephen-nordic**
Fixes #7

### CODE REVIEW

**Comment by torsteingrindvik** [comment]
Windows fails due to a Unix dependency.
That should be fixed by using attributes and/or feature flags where necessary.

Do you want to

1. Add a commit with the necessary fixes,
2. Get a more thorough description of what needs to be done then add the commit, or
3. I'll add the commit

?

**Comment by torsteingrindvik** [comment]
Also, if in your PR description you write

> Fixes #22 

Then that issue will be automatically closed when this PR is merged.

**Comment by stephen-nordic** [comment]
> Windows fails due to a Unix dependency. That should be fixed by using attributes and/or feature flags where necessary.
> 
> Do you want to
> 
> 1. Add a commit with the necessary fixes,
> 2. Get a more thorough description of what needs to be done then add the commit, or
> 3. I'll add the commit
> 
> ?

We'll do this and see if it fixes it? However the /exit endpoint is still unimplemented for platforms other than Linux

```rust
#[cfg(unix)]
use tokio::signal::unix::{signal, SignalKind};

#[cfg(windows)]
use tokio::signal::windows::{signal, SignalKind};

```

**Comment by stephen-nordic** [comment]
> > Windows fails due to a Unix dependency. That should be fixed by using attributes and/or feature flags where necessary.
> > Do you want to
> > 
> > 1. Add a commit with the necessary fixes,
> > 2. Get a more thorough description of what needs to be done then add the commit, or
> > 3. I'll add the commit
> > 
> > ?
> 
> We'll do this and see if it fixes it? However the /exit endpoint is still unimplemented for platforms other than Linux
> 
> ```rust
> #[cfg(unix)]
> use tokio::signal::unix::{signal, SignalKind};
> 
> #[cfg(windows)]
> use tokio::signal::windows::{signal, SignalKind};
> ```

So there's no 'signal' for windows. @torsteingrindvik What do you think is the most appropriate way to fix this?

**Comment by torsteingrindvik** [comment]
Yep, there is not cross-platform straightforward way to do it.
Could you try to basically make the code paths which now use signals just be a no-op on Windows?

Sort of like what `exit_server()` is for Windows now.

Later we can follow up by not using an OS signal for exiting, but rather something like a one shot channel from tokio.

**Comment by stephen-nordic** [comment]
Ready for review now :)

File: .github/workflows/basics.yml
Comment (torsteingrindvik) [inline .github/workflows/basics.yml:1]: We should run clippy as well.

File: .github/workflows/basics.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      matrix:
+        os: [windows-latest, ubuntu-latest, macos-latest]
+    runs-on: ${{ matrix.os }}
+    timeout-minutes: 30
+    steps:
+      - uses: actions/checkout@v3
+      - name: cargo build
+        run: cargo build --bin serial-keel
+      - name: cargo test
+        run: cargo test
```
Comment (torsteingrindvik) [inline .github/workflows/basics.yml]: This should have `--all-features`.
Once again #11 is a bother :)

File: README.md [hunk 2]
```diff
@@ -1,4 +1,4 @@
-
+![Serial Keel Build](https://github.com/torsteingrindvik/serial-keel/.github/workflows/basics.yml/badge.svg)
```
Comment (stephen-nordic) [inline README.md:1]: We'll know for sure whether this works only after merge :(

Comment (torsteingrindvik) [reply README.md:1]: That's ok. I like the idea though :slightly_smiling_face: 
If it fails for some reason after merge we can just force push that commit in `main` until it works.

File: .github/workflows/basics.yml (hunk 1 above)
Comment (stephen-nordic) [reply .github/workflows/basics.yml]: done

File: .github/workflows/basics.yml
Comment (stephen-nordic) [reply .github/workflows/basics.yml:1]: also done

**Comment by torsteingrindvik** [comment]
I noticed the docs job fails:

```
warning: unresolved link to `Endpointid`
  --> core/src/endpoint.rs:99:28
   |
99 | /// which should look at [`Endpointid`] instead.
   |                            ^^^^^^^^^^ no item named `Endpointid` in scope
   |
   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
   = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default
```

could you look into how `cargo docs` can be run such that warnings are errors?

**Comment by torsteingrindvik** [comment]
Also when running the tests, we get log output. That's a bit too noisy.
That's because I have littered some tests with:

```rust
serial_keel::logging::init().await;
```

could you search for those lines within the tests and delete them?

File: .github/workflows/basics.yml [hunk 3]
```diff
@@ -34,7 +34,7 @@ jobs:
     steps:
       - uses: actions/checkout@v3
       - name: cargo doc
-        run: cargo doc --no-deps
+        run: /bin/bash -c "cargo doc 2>&1 > /dev/null | grep -q "warning:" ; test $? -ne 0"
```
Comment (torsteingrindvik) [inline .github/workflows/basics.yml]: https://stackoverflow.com/questions/73981385/make-cargo-doc-fail-if-there-are-warnings

Could you try `RUSTDOCFLAGS` as suggested here?

Comment (stephen-nordic) [reply .github/workflows/basics.yml]: yeah, I did see that. It didn't seem to work. `cargo doc` still returned exit code 0 with warnings. Maybe I need to try more. I will check back again tomorrow

Comment (stephen-nordic) [reply .github/workflows/basics.yml]: It works now. I just mustn't have set the environment variable correctly. Resolving this conversation.

**Comment by stephen-nordic** [comment]
> Also when running the tests, we get log output. That's a bit too noisy. That's because I have littered some tests with:
> 
> ```rust
> serial_keel::logging::init().await;
> ```
> 
> could you search for those lines within the tests and delete them?

done

**Review by torsteingrindvik — APPROVED** [review]
Great job, this is very valuable to have going forward. Especially cool to see that GitHub CI runners are using Serial Keel live in tests :slightly_smiling_face: 

There is a slight hiccup in that we likely should have had `--mocks-share-endpoints` or `--all-features` to all `build`, `test`, `doc`, `clippy` invocations, but let's not think much about it now.
It's been too much of a nuisance, we can make the default better in #11 .

**Pull request merged**